random: Add syscalls for random subsystem

ceolin opened the pull request:
Create syscalls to make possible using random APIs from user mode
threads. These APIs can have different implementations, like using
entropy driver or Xoroshiro128. Some of these implementations also have
some globals to preserve state between calls.

Make it run entire in user space would require user adding these globals
to their memeory domains and/or grant access to entropy device. Syscalls
simplify its usage.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Fixes #26499

Commit:
random: Add syscalls for random subsystem

Create syscalls to make possible using random APIs from user mode
threads. These APIs can have different implementations, like using
entropy driver or Xoroshiro128. Some of these implementations also have
some globals to preserve state between calls.

Make it run entire in user space would require user adding these globals
to their memeory domains and/or grant access to entropy device. Syscalls
simplify its usage.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

d3zd3z reviewed (approved):
I think I prefer this over the #25890 solution. As I understand, this will keep the DBRG state away from userspace.

d3zd3z commented:
A little reasoning as to why I prefer this over the #25890 solution.

1. This better protects the DBRG state. The internal state of the DBRG needs to be protected the same as a private or encryption key. Making these syscalls allows the state to be kept entirely in kernel.
2. With hardware crypto, or something like TF-M, the state will need to be kept on the secure side, and we would need syscalls to access it anyway.

jukkar reviewed (approved):
I like this one better too, +1 from me.

jukkar commented:
Could we get this merged soon, I would need this support in other PRs.

dleach02 commented:
Is there a reason this set of changes are not including rand32_ctr_drbg.c or am I missing something here?

ceolin commented:
> Is there a reason this set of changes are not including rand32_ctr_drbg.c or am I missing something here?

nope, I just missed, I remembering doing it though. Will add it though

Commit:
random: Include header where it is used

Unit tests were failing to build because random header was included by
kernel_includes.h. The problem is that rand32.h includes a generated
file that is either not generated or not included when building unit
tests. Also, it is better to limit the scope of this file to where it is
used.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

dleach02 commented on include/kernel_includes.h:
```diff
@@ -33,7 +33,6 @@
 #include <mempool_sys.h>
 #endif
 #include <kernel_version.h>
-#include <random/rand32.h>
```
why are you taking the header include out of kernel_includes?

dleach02 replied:
nevermind, I looked at the commit message where it is described... I suspect you ran into this when adding the ctrdbg stuff... ;-)

ceolin replied:
nope, the initial patch had this problem. rand32.h now includes an auto-generated file that is not available for unit tests.

ceolin commented:
@nashif @jukkar can you restart CI, I'm not seeing any issue there just some some error cause by timeout.

pfalcon commented:
> just some some error cause by timeout.

This happens more and more often, for more and more people and cases. Turns out, Zephyr's test runner is designed for hanging, then timing out: https://github.com/zephyrproject-rtos/zephyr/issues/26209 . And when reported, the best response received is that it "had been such for years".

> can you restart CI 

Just close/reopen, that's the modern version of the older "recheck" trick. (I did that a few hours ago. As you can see, it still fails. Previously are suspected non-reliable hardware used for tests. Now I suspect that just test runner has erratic behavior, and it just gets exposed.)

PerMac commented:
> > just some some error cause by timeout.
> 
> This happens more and more often, for more and more people and cases. Turns out, Zephyr's test runner is designed for hanging, then timing out: #26209 . And when reported, the best response received is that it "had been such for years".

The testing framework is currently undergoing some greater verification and update. You are right that some problems were not tackled for too long but it was not because of bad will of the framework maintainers but due to the nature of an opensource project. I believe we were unaware of many problems in the design. We are now consolidating our efforts to improve the performance of the framework. 
One of the issues with timeout behavior is being addressed by #26223. We are open to changes and highly value all the input that can lead to an improvement of the framework, so please don't hesitate to raise more issues so we can be aware of pitfalls that are certainly still in the framework.

ceolin commented:
> @PerMac, can you tell though where this is failing in shippable? I looking at the failing task but can't tell what the problem is? And given the "turn it off turn it on" that @pfalcon performed didn't seem to fix it I'm worried that there is something else going on here.

Yep, that would be fantastic

PerMac commented:
I am not sure if this won't be just a noise, but I see failing tests in the build. I found them by looking in buildkite (new Ci tested) checks. Stage 19/20 failed. On the log tab you can download full log and see errors there e.g. 
```
FAILED: CMakeFiles/app.dir/src/main.c.obj 

ccache /usr/bin/gcc -DBUILD_VERSION=zephyr-v2.3.0-353-gb87a0903622b -DKERNEL -D_FORTIFY_SOURCE=2 -D_POSIX_C_SOURCE=200809 -D_XOPEN_SOURCE=600 -D_XOPEN_SOURCE_EXTENDED -D__ZEPHYR__=1 -I/workdir/zephyr/subsys/net/ip -I/workdir/zephyr/include -Izephyr/include/generated -I/workdir/zephyr/soc/posix/inf_clock -I/workdir/zephyr/boards/posix/native_posix -I/workdir/zephyr/subsys/testsuite/include -I/workdir/zephyr/subsys/testsuite/ztest/include -I/workdir/zephyr/subsys/net/l2 -Os -imacros/workdir/zephyr/sanity-out/native_posix/tests/net/traffic_class/net.traffic_class.tx_8_rx_7_sr_b/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -imacros/workdir/zephyr/include/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-address-of-packed-member -Wno-pointer-sign -Wpointer-arith -Wno-unused-but-set-variable -Werror=implicit-int -Werror -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-strict-overflow -fno-reorder-functions -fno-defer-pop -fmacro-prefix-map=/workdir/zephyr/tests/net/traffic_class=CMAKE_SOURCE_DIR -fmacro-prefix-map=/workdir/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/workdir=WEST_TOPDIR -ffunction-sections -fdata-sections -m32 -include /workdir/zephyr/arch/posix/include/posix_cheats.h -fno-freestanding -std=c11 -MD -MT CMakeFiles/app.dir/src/main.c.obj -MF CMakeFiles/app.dir/src/main.c.obj.d -o CMakeFiles/app.dir/src/main.c.obj   -c /workdir/zephyr/tests/net/traffic_class/src/main.c

/workdir/zephyr/tests/net/traffic_class/src/main.c: In function ‘generate_mac’:

/workdir/zephyr/tests/net/traffic_class/src/main.c:252:16: error: implicit declaration of function ‘sys_rand32_get’ [-Werror=implicit-function-declaration]

  252 |  mac_addr[5] = sys_rand32_get();

      |                ^~~~~~~~~~~~~~

cc1: all warnings being treated as errors

[17/114] Building C object zephyr/CMakeFiles/zephyr.dir/lib/os/notify.c.obj

[18/114] Building C object zephyr/CMakeFiles/zephyr.dir/lib/os/mempool.c.obj

ninja: build stopped: subcommand failed.
```
You can go in that stage to the artifacts tab and download the XML report. It shows 57 failures, I think the most in `native_posix:tests/net/traffic_class/`

dleach02 commented:
@PerMac, This is a real bug. It is a warning that the sys_rand32_get is undefined. @ceolin modified the header files and much of the changes in this PR involve added the rand header file to those c files that need it. It appears that tests/net/traffic_class/src/main.c is one of them.

But this highlights a couple of problems with CI:

1. We have reached a point where the sporadic errors have conditioned us to first restart CI before actually digging into the logs. We really need to address the sporadic errors.
2. It would be useful to have this error/warning be displayed in the summary screen instead of having to dig through a lengthy log file.

PerMac commented:
@dleach02 Glad that I was able to help you. I agree with your comments. I think it might be some bug in the CI that the test results are not shown correctly and can only be seen when digging in logs. I will raise this issue at #infrastructure channel on slack. I know that right now @galak and @nashif are working on improving the CI and I will ask them to look at this.  I think you might also raise a git issue describing the problem so you can track if there is progress.

dleach02 commented:
I've been viewing this from Shippable perspective because I been thinking the buildkite is still experimental. Since Shippable is going away it probably isn't worth trying to get the information to display there of the warnings. From buildkite it looks like you only can see things by opening up the log file? Not sure as I haven't been looking at buildkit output yet.

But for the random CI failures, there are already a couple of issue bugs entered so I'm not sure what to put into a new issue bug. But I understand the sentiment.

ceolin commented:
> @PerMac, This is a real bug. It is a warning that the sys_rand32_get is undefined. @ceolin modified the header files and much of the changes in this PR involve added the rand header file to those c files that need it. It appears that tests/net/traffic_class/src/main.c is one of them.
> 
> But this highlights a couple of problems with CI:
> 
>     1. We have reached a point where the sporadic errors have conditioned us to first restart CI before actually digging into the logs. We really need to address the sporadic errors.
> 
>     2. It would be useful to have this error/warning be displayed in the summary screen instead of having to dig through a lengthy log file.

that is really an issue, have fixed the problem and submitted a new version. Thanks !

dleach02 commented on subsys/random/rand32_handlers.c:
```diff
@@ -0,0 +1,33 @@
+
+static inline void z_vrfy_sys_rand_get(void *dst, size_t len)
+{
+	Z_OOPS(Z_SYSCALL_MEMORY_WRITE(dst, len));
```
Are you suggesting making a copy of the value in dst? How does this value get changed between the Z_SYSCALL_MEMORY_WRITE() and z_impl_sys_rand_get()? Now if it was a pointer to a pointer then there might be a problem but the initial call to is the memory area to validate.

ceolin commented:
@nashif The CI error seems not to be related with this patch, I've checked the test that has failed locally and it is working.
